### ISSUE TITLE
ARCHBOM-2010: How does Event Bus interact with OpenedX Django Signals?

### ISSUE THREAD

**Issue by jinder1s** [eduNEXT/openedx-events#39]
### Context
The design intent behind eduNEXT/openedx-events was to be able to reuse the event definitions for both internal events (signals) and external events (message bus).

Currently, the plan seems to have all events be signals. Some of these signals would be received by a Kafka producer, which will then publish the data in the signal to the event bus. The Kafka consumer would then listen to the event and convert it back to a signal.

This would mean:
- no event would ever just be a kafka event. For every Kafka event, there would be a signal event (not vice-versa).
- Kafka event is only produced if a signal is send (by sender) and received (by Kafka Producer).

The definitions for the django signals are defined in [signal.py](https://github.com/eduNEXT/openedx-events/blob/main/openedx_events/learning/signals.py#L18). Specifically, django signals definition in encapsulated in OpenEdxPublicSignal class. OpenEdxPublicSignal is a django signal classs. 

### Questions
1. How sure are we that we want a signal event for every Kafka event?
2. Does OpenEdxPublicSignal actually need to be a signal? Even if we decide to start with Event Bus events always being sent downstream of the signal (for now), does it make sense to enforce this in the class design and leave no room for adjustments?
     This class overrides some signal methods to get users to use a special method.

     a. Is it possible for this class to just contain a signal class instead (when needed), or can the signal class not be dynamic?
3. Could OpenEdxPublicSignal be responsible for sending both signal and Kafka event bus events?
    This complexity may not be worth it now without a strong need, but it could be noted as a possibility. For now, maybe this should be documented as a rejected alternative, simply because it isn’t yet required.

4. Could we want the design of an event to look different when internal vs external? For example, how will we design these lifecycle events? See  [ARCHBOM-2008](https://openedx.atlassian.net/browse/ARCHBOM-2008)

**Comment by ormsbee** [comment]
I'm sympathetic to the concern that the present design is too strongly tied to Signals. It's a tradeoff to try to reduce the cognitive load of the feature by piggy-backing on a commonly used Django mechanism, and it's clearly not without drawbacks. That being said, I would really appreciate more concrete use cases and scenarios for which it would be problematic. I worry that trying to have this conversation without concrete examples to ground the discussion will lead us down a rabbit hole of hypotheticals which are hard to weigh against each other.

> 1. How sure are we that we want a signal event for every Kafka event?

I can't be certain, of course. The only plausible scenario I can currently think of where we'd want to send something via Kafka and not a local signal is if we have a service that explicitly discourages plugin or code-level integrations and wants to force all integration to happen over message queues. But in that case, having the Signals doesn't really hurt the project–it just wouldn't really be used.

I suppose there might be some Kafka bookkeeping / service heartbeat type of messages that get sent? Though that feels like a very different thing from the type of events defined in this repo.

Some points in favor of Kafka events being a subset of Signals sent:

* Signals are much cheaper to send, and offer pretty much no additional work as compared to sending via Kafka.
* Signals are already widely used in edx-platform and elsewhere–we know that people will use them.
* Signals don't require extra infrastructure and are simpler for people to test with.

That being said, if we do later find that we need to send events only to the message queue and not to Signals, it's possible we just use a different mechanism for them. OpenEdxPublicSignal could be the class we use when it sends to both. I'm still not clear on what that a concrete use case for this would be though.

> Does OpenEdxPublicSignal actually need to be a signal? Even if we decide to start with Event Bus events always being sent downstream of the signal (for now), does it make sense to enforce this in the class design and leave no room for adjustments?
> This class overrides some signal methods to get users to use a special method.
> 
> a. Is it possible for this class to just contain a signal class instead (when needed), or can the signal class not be dynamic?

We could. The constructor for this class could just as easily create an instance of a Signal under the covers. It's just that if we did that, we'd have to expose that underlying signal when hooking it up using Django's `@receiver`, so it'd look like:

```python
@receiver(STUDENT_REGISTRATION_COMPLETED.signal)  # Note the .signal here
def send_welcome_notification(sender, user, **kwargs):
    # do something here
```

But I think that actually makes it less clear in terms of what operations you're supposed to do on the event vs. the signal object there. We don't want people to ever use the signal's `send/send_robust` methods, because we want to force them to use our `send_event` that has validation and metadata wrapping.

We could create a signal object internally and proxy the relevant calls from OpenEdxPublicSignal to that Signal instance, but that's an option we always have open to us anyway. If we think the external API of OpenEdxPublicSignal would look the same, we can shift around the question of subclass vs. proxy-to-signal as an implementation detail later on, if that becomes beneficial.

We could completely roll our own event handling that hooks in-process notification and triggers both signals and Kafka messages (so treating Signals as a peer to Kafka and other message-queue-bound events). But proper dispatch for that sort of thing is non-trivial code that has to contend with multithreading and memory leaks, and we're probably not going to do better than battle-tested Django signal code.

I don't think any of those scenarios is more compelling than just subclassing Signal for the moment.

> 3. Could OpenEdxPublicSignal be responsible for sending both signal and Kafka event bus events?
> This complexity may not be worth it now without a strong need, but it could be noted as a possibility. For now, maybe this should be documented as a rejected alternative, simply because it isn’t yet required.

Sure. At the end of the day, the interface for sending is just a method call to `send_event`. We could have some sort of flag in the constructor for OpenEdxPublicSignal that passes along a flag to listeners that determines whether it should get passed to Kafka or not (assuming again that Kafka signals are a subset of in-process signals). We could also have something in the `send_event` that explicitly sends out to Kafka (or a more general abstraction)–we'll need some variant of this route if we want Kafka connection failures to fail the entire request.

I think that's worth noting, but not worth implementing until we have a concrete use case for it. I suspect that any such requirement will come with other complications.

One thing I like about implementing Kafka-message sending as an app listening for signals is that it makes it easy to implement other transports in parallel. As much as possible, I would like to avoid having Kafka baked into openedx-events, and allow for someone to write an adapter for Redis Streams without too much difficulty (which has already come up in earlier discussions). Signals are the de facto mechanism for this type of notification in the Django Framework, so I don't mind coupling there as much. Kafka to me is a transport that should ideally be swappable.

> 4. Could we want the design of an event to look different when internal vs external? For example, how will we design these lifecycle events? See ARCHBOM-2008

The event being sent externally is always going to be more constrained. Size, serializability, partitioning, ordering–those are all much bigger problems going over the network than they are in-process. At the same time, I think that those constraints aren't necessarily bad. The small size and simplicity that makes messages practical to send over a message queue also make it easy and predictable for extensions authors to work with a signal.

So I think we should try to have them be the same if we can.

**Comment by ormsbee** [comment]
FYI @mariajgrimaldi, @felipemontoya, @feanil

**Comment by mariajgrimaldi** [comment]
Hi there! Thank you both again for the detailed description of the matter. 

### Answers (from the info I've got & just if relevant)

> 3. Could OpenEdxPublicSignal be responsible for sending both signal and Kafka event bus events?
>     This complexity may not be worth it now without a strong need, but it could be noted as a possibility. For now, maybe this should be documented as a rejected alternative, simply because it isn’t yet required.

Picking up what Dave suggested about _implementing Kafka-message sending as an app listening for signals_, it could be interesting to think about an adapter application for general use like `openedx-events-over-the-wire`, so redis or another technology could be used (just an idea, I don't have much experience on this matter).

**Comment by feanil** [comment]
@jinder1s thanks for the thoughtful write-up and questions.  A lot of them seem to be dealing with new issues we might face in the future if we go down the current path of signals first, then publish to external sources(event-bus) off of the signal.  Based on Dave's response, and the fairly small interface that the `OpenEdxPublicSignal` class has, it feels like we are not actually locking our selves in to anything at the moment.

Is there an issue today that we're running into that I don't understand or is there a specific use-case we have today that we can't meet with the strategy laid out above?

**Comment by robrap** [comment]
Hello all. Thank you for your thoughts. They are quite helpful.

I feared the current design might be overly constrained, but don't have enough experience with it to know how and if it will really be an issue. Others seem to believe this isn't as reversible as I'd imagined, so maybe I just need to not sweat it.

Here are some open questions I still have:
1. I wonder about observability. I'm hoping that layering signals and event bus events, both of which can be difficult to trace, don't make debugging or understanding the system more difficult.
2. I'm not yet sure what needs we have that are specific to external events. For example, designating the partitioning key, or handling an event ID that can be used for observability.
3. Other questions I have around consuming messages and schema evolution I'll address in https://github.com/eduNEXT/openedx-events/issues/38.

Thanks again.

**Comment by felipemontoya** [comment]
I was catching up on this discussion and a different angle came to my mind.

One of the main goals we had when we started sketching up the foundations for this events work, was to provide a certain ammount of stability for developers that wanted to extend the functionality of the lms/cms services. Moving data using an event bus and having multiple teams consume the different events is more of the same problem, right? I could imagine that some teams will end up using django for their consumser services and thus will have the same issues we as community members faced when doing plugins.

- How do I keep my consumer working when someone adds more data to an event I'm listening?
- Am I bound to upgrade to always the latest version of an event or will the one I consume be deprecated timely?
- How do I test my code and make sure that changes to the events break in CI instead of prod?

Does kafka address this differently? I'd like to think that the robustness of the event definions in an external library and all the rigidity brought by the attr classes could also help event bus consumers. Even if they are not necessarily using django and can import this library and somehow connect the kafka events back into signals.

**Comment by robrap** [comment]
It seems like everyone is pretty confident in the proposed design, where we lead with signals (and potentially send signals on the consumer side). We'll need to make more progress to verify if there are any hidden showstoppers.

On the observability front, I'm not sure what is waiting for us, but we'll see more when we get there.

Additionally, as I noted, we may need to add to the signal definition with event-bus only fields like the partitioning key. This isn't really a blocker, but just something we'll need to figure out.

@ormsbee: Thank you again for all your thoughts above. I think I imagined something more like an `OpenEdxPublicEvent` with your `proxy-to-signal` design, but I agree that this may not be a necessary change. It is just more closely aligned to what I would have pictured as an event definition, and potentially could have made it more clear if we had more non-signal use cases in the future. But, I agree that that is a big fuzzy _if_ at this point, so I'm ready to move on from that design for now.

@felipemontoya: I'm sorry if my response was that misleading. In general, I agree that there are very similar concerns with the event bus. See https://github.com/eduNEXT/openedx-events/issues/38 for a more detailed discussion of how the attrs classes are being used for schema evolution, and some open questions around that.

**Comment by ormsbee** [comment]
@robrap, @jinder1s: Thanks for hashing all of this out. I have a feeling that we're going to be referring back to this issue and #38 repeatedly in the coming months. 😄 

Have we resolved the general questions that were originally posed here? Should we close this issue and open up more specific ones as they come up? I suspect that the partitioning question alone will be big enough for its own Issue. 😛

**Comment by robrap** [comment]
I see ADR(s) in our near future, but I think my questions have settled for now. 
I know @jinder1s may need a little more time with this, but not sure if he has any further questions. Since he opened the issue, maybe he can close it if and when he is ready.

**Comment by jinder1s** [comment]
Hi, thanks everyone for the great responses. For me, this was a great exercise in getting the right context to move forward with this work. I'm going to close this issue cause it has helped me with getting the context I need. I'm currently in the process of trying to figure out what ADRs need to be written to better document this. So look forward to being pinged on PRs for those ADRs. Thank You!
